ouroboros-network/src/Ouroboros/Network/BlockFetch/Decision/BulkSync.hs

```diff
@@ -0,0 +1,605 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+-- | BulkSync decision logic
+--
+-- This module contains the part of the block fetch decisions process that is
+-- specific to the bulk sync mode. This logic reuses parts of the logic for the
+-- deadline mode, but it is inherently different.
+--
+-- Natural language specification
+-- ------------------------------
+--
+-- Definitions:
+--
+-- - Let @inflight :: peer -> Set blk@ be the outstanding blocks, those that
+--   have been requested and are expected to arrive but have not yet.
+--
+-- - Let @peersOrder@ be an order of preference among the peers. This order is
+--   not set in stone and will evolve as we go.
+--
+-- - Let @currentPeer :: Maybe peer@ be the “current peer” with which we are
+--   interacting. If it exists, this peer must be the best according to
+--   @peersOrder@, and the last fetch request must have been sent to them.
```

Comment:
Actually, now that I see your comment, I don't think it is necessary for `currentPeer` to be the best according to `peersOrder`, since we will not look at `peersOrder` until this peer has starved us or we have no in-flight request with this peer anymore, probably because it doesn't have any blocks for us anymore.